### ISSUE TITLE
MethodError: no method matching zero(::Type{Variable})

### ISSUE THREAD

**Issue by Mark-Bowers** [tkoolen/QPWalkingControl.jl#4]
```julia
Notebook: Atlas walking: Error During Test at REPL[182]:15
  Got exception outside of a @test
  LoadError: MethodError: no method matching zero(::Type{Variable})
  Closest candidates are:
    zero(!Matched::Type{Missing}) at missing.jl:82
    zero(!Matched::Type{LibGit2.GitHash}) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.2\LibGit2\src\oid.jl:220
    zero(!Matched::Type{Pkg.Resolve.VersionWeights.VersionWeight}) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.2\Pkg\src\resolve\VersionWeights.jl:19
    ...
  Stacktrace:
   [1] zero(::Variable) at C:\Users\mark_\.julia\packages\Parametron\ihvFk\src\functions.jl:640
   [2] _vecdot at C:\Users\mark_\.julia\packages\StaticArrays\3KEjZ\src\linalg.jl:227 [inlined]
   [3] dot at C:\Users\mark_\.julia\packages\StaticArrays\3KEjZ\src\linalg.jl:217 [inlined]
   [4] dot(::FreeVector3D{SArray{Tuple{3},Variable,1,3}}, ::FreeVector3D{SArray{Tuple{3},Variable,1,3}}) at C:\Users\mark_\.julia\packages\RigidBodyDynamics\Oe3ru\src\spatial\threevectors.jl:75
   [5] macro expansion at C:\Users\mark_\.julia\packages\Parametron\ihvFk\src\lazyexpression.jl:59 [inlined]
   [6] evalexpr at C:\Users\mark_\.julia\packages\Parametron\ihvFk\src\lazyexpression.jl:56 [inlined]
   [7] (::Parametron.LazyExpression{typeof(dot),Tuple{FreeVector3D{SArray{Tuple{3},Variable,1,3}},FreeVector3D{SArray{Tuple{3},Variable,1,3}}}})() at C:\Users\mark_\.julia\packages\Parametron\ihvFk\src\lazyexpression.jl:53
   [8] optimize_toplevel(::Parametron.LazyExpression) at C:\Users\mark_\.julia\packages\Parametron\ihvFk\src\lazyexpression.jl:191
   [9] objectiveterm(::QPControl.ContactPoint{4}, ::Model{Float64,Optimizer}) at C:\Users\mark_\.julia\packages\QPControl\ohfic\src\contacts.jl:78
   [10] addcontact!(::MomentumBasedController{4,Optimizer,MechanismState{Float64,Float64,Float64,TypeSortedCollections.TypeSortedCollection{Tuple{Array{Joint{Float64,QuaternionFloating{Float64}},1},Array{Joint{Float64,Revolute{Float64}},1}},2}}}, ::RigidBody{Float64}, ::QPControl.ContactPoint{4}) at C:\Users\mark_\.julia\packages\QPControl\ohfic\src\lowlevel\momentum.jl:137
   [11] addcontact!(::MomentumBasedController{4,Optimizer,MechanismState{Float64,Float64,Float64,TypeSortedCollections.TypeSortedCollection{Tuple{Array{Joint{Float64,QuaternionFloating{Float64}},1},Array{Joint{Float64,Revolute{Float64}},1}},2}}}, ::RigidBody{Float64}, ::Point3D{SArray{Tuple{3},Float64,1,3}}, ::FreeVector3D{SArray{Tuple{3},Float64,1,3}}, ::Float64) at C:\Users\mark_\.julia\packages\QPControl\ohfic\src\lowlevel\momentum.jl:147
   [12] top-level scope at C:\Users\mark_\.julia\packages\QPWalkingControl\7CP7N\notebooks\Atlas walking.ipynb:In[+4]:16
   [13] include_string(::Module, ::String, ::String) at .\loading.jl:1064
   [14] my_include_string(::Module, ::String, ::String, ::Nothing, ::Bool) at C:\Users\mark_\.julia\packages\NBInclude\m4rfj\src\NBInclude.jl:29
   [15] #nbinclude#1(::Bool, ::UnitRange{Int64}, ::Regex, ::typeof(identity), ::Bool, ::typeof(NBInclude.nbinclude), ::Module, ::String) at C:\Users\mark_\.julia\packages\NBInclude\m4rfj\src\NBInclude.jl:82
   [16] (::getfield(NBInclude, Symbol("#kw##nbinclude")))(::NamedTuple{(:regex,),Tuple{Regex}}, ::typeof(NBInclude.nbinclude), ::Module, ::String) at .\none:0
   [17] top-level scope at REPL[182]:17
   [18] top-level scope at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.2\Test\src\Test.jl:1113
   [19] top-level scope at REPL[182]:17
   [20] eval(::Module, ::Any) at .\boot.jl:330
   [21] top-level scope at REPL[182]:12
   [22] eval(::Module, ::Any) at .\boot.jl:330
   [23] eval_user_input(::Any, ::REPL.REPLBackend) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.2\REPL\src\REPL.jl:86
   [24] macro expansion at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.2\REPL\src\REPL.jl:118 [inlined]
   [25] (::getfield(REPL, Symbol("##26#27")){REPL.REPLBackend})() at .\task.jl:268
  in expression starting at C:\Users\mark_\.julia\packages\QPWalkingControl\7CP7N\notebooks\Atlas walking.ipynb:In[+4]:2

Test Summary:           | Error  Total
Notebook: Atlas walking |     1      1
ERROR: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
```

**Comment by tkoolen** [comment]
This was fixed in https://github.com/tkoolen/Parametron.jl/pull/106, released in v0.9.1 of Parametron. Perhaps you need to do a Pkg update?

**Comment by Mark-Bowers** [comment]
Twan, Thank you for your assistance, and thank you for creating all these robotics packages!

**Comment by tkoolen** [comment]
Just to confirm, did the update fix the issue?

**Comment by Mark-Bowers** [comment]
I had to wrestle with the package manager a bit, but was finally able to get it to use the right version, and get past this issue. The package manager was not able to resolve the package name Parametron when I first attempted to add it. This may be because the registry was out of date, or I may have misspelled it (The package manager seems to do an automatic registry update if you give it a package name that it doesn't recognize, so I must have not provided the name correctly). As a result, I added it using the git repo and was on the master branch (0.9.1), and still don't know how I ended up with an older version.

Once I verified that Parametron was in the registry, I removed, and then re-added it. That resulted in it fetching version 0.8. I tried to force it to add 0.9.1, but other packages' compatibility requirements prevented it. I ended up removing 4 packages before I was able to get the latest version of Parametron, but weirdly was able to then re-add those 4 packages without incident. Amazingly, that still didn't work. After verifying that I was getting the right version (pathof(Parametron)), I deleted the compiled package (4wlnK.ji), restarted Julia, and was able to get it to compile the correct version.

I was able to get significantly further, but am now running into an issue in the final steps of working through the example in the notebook:

```julia
julia> Animation(mvis, sol)
ERROR: MethodError: Cannot `convert` an object of type MechanismVisualizer{MechanismState{Float64,Float64,Float64,TypeSortedCollections.TypeSortedCollection{Tuple{Array{Joint{Float64,QuaternionFloating{Float64}},1},Array{Joint{Float64,Revolute{Float64}},1}},2}},Visualizer} to an object of type Dict{MeshCat.SceneTrees.Path,MeshCat.AnimationClip}

I am just starting to investigate what may be causing this, but here is the current status of my installed packages. My first guess is that I still have versioning issues:

(v1.2) pkg> status
    Status `C:\Users\mark_\.julia\environments\v1.2\Project.toml`
  [537997a7] AbstractPlotting v0.9.10
  [436ed74b] AtlasRobot v0.4.0
  [6e4b80f9] BenchmarkTools v0.4.3
  [9e28174c] BinDeps v0.8.10
  [ad839575] Blink v0.10.1
  [5ae59095] Colors v0.9.6
  [8f4d0f93] Conda v1.3.0
  [a6e380b2] ControlSystems v0.5.3
  [150eb455] CoordinateTransformations v0.5.0
  [a26e6606] DICOM v0.3.1
  [2b5f629d] DiffEqBase v5.20.1
  [459566f4] DiffEqCallbacks v2.8.0
  [5789e2e9] FileIO v1.0.7
  [f6369f11] ForwardDiff v0.10.3
  [f7f18e0c] GLFW v3.1.0
  [4d00f742] GeometryTypes v0.7.6
  [7073ff75] IJulia v1.20.0
  [d3863d7c] InteractBase v0.10.3
  [97c1335a] JSExpr v0.5.0
  [ee78f7c6] Makie v0.9.3
  [b8f27783] MathOptInterface v0.8.4
  [931e9471] MechanismGeometries v0.4.1
  [283c5d60] MeshCat v0.7.0
  [6ad125db] MeshCatMechanisms v0.5.0
  [7269a6da] MeshIO v0.3.1
  [e6723b4c] Meshing v0.4.1
  [7528ae0c] MuJoCo v0.1.0 #master (https://github.com/klowrey/MuJoCo.jl)
  [9ecbcd48] MujocoSim v0.1.0 #master (https://github.com/klowrey/MujocoSim.jl)
  [a975b10e] Mux v0.7.0
  [0db19996] NBInclude v2.1.0
  [ab2f91bb] OSQP v0.5.2
  [1dea7af3] OrdinaryDiffEq v5.15.1
  [d96e819e] Parameters v0.11.0
  [41fac6d6] Parametron v0.9.1
  [145d500b] PlanarConvexHulls v0.3.0
  [91a5bcdd] Plots v0.26.2
  [438e738f] PyCall v1.91.2
  [d1d18edd] QPControl v0.1.0 #master (https://github.com/tkoolen/QPControl.jl)
  [0b169a85] QPWalkingControl v0.1.0 #master (https://github.com/tkoolen/QPWalkingControl.jl)
  [295af30f] Revise v2.1.10
  [366cf18f] RigidBodyDynamics v2.1.0
  [e61f16d8] RigidBodySim v1.3.0
  [6038ab10] Rotations v0.11.1
  [276daf66] SpecialFunctions v0.8.0
  [90137ffa] StaticArrays v0.11.0
  [6f584044] StaticUnivariatePolynomials v0.6.0
  [2913bbd2] StatsBase v0.32.0
  [c74b26e8] ValkyrieRobot v0.2.0
  [0f1e0344] WebIO v0.7.0
```

Incidentally, In working through the notebook example at the command line, I found that I had to add a 'using Parametron' to the example.

**Comment by tkoolen** [comment]
The `Animation(mvis, sol)` error is because that requires https://github.com/JuliaRobotics/MeshCatMechanisms.jl/pull/34. As this package is unreleased and under development, I currently have a `Manifest.toml` in the repo with specific versions of dependencies, some of which refer to commits that have not made it into a release. If you activate QPWalkingControl's own environment and instantiate it, you should get dependency versions that are compatible. But if you add QPWalkingControl to a different environment, the `Manifest.toml` versions are ignored.

Sorry that this package is not in the most polished state right now. I'm working on finishing up my thesis, after which I'd like to clean things up and add it to the package registry.

**Comment by Mark-Bowers** [comment]
Instantiating environments is new to me, and I didn't have much luck attempting to set one up for QPWalkingControl. For the record, I tried:

```julia
(v1.2) pkg> activate QPWalkingControl
Activating new environment at `C:\Users\mark_\AppData\Local\Julia-1.2.0\QPWalkingControl\Project.toml`

(QPWalkingControl) pkg> status
    Status `C:\Users\mark_\AppData\Local\Julia-1.2.0\QPWalkingControl\Project.toml`
  (empty environment)

julia> dirname(pathof(QPWalkingControl))
"C:\\Users\\mark_\\.julia\\packages\\QPWalkingControl\\7CP7N\\src"

julia> cd("C:\\Users\\mark_\\.julia\\packages\\QPWalkingControl\\7CP7N\\src\\")

julia> using Pkg

julia> @__DIR__
"C:\\Users\\mark_\\.julia\\packages\\QPWalkingControl\\7CP7N\\src"

julia> using Pkg

julia> Pkg.activate(joinpath(@__DIR__, ".."))
Activating environment at `C:\Users\mark_\.julia\packages\QPWalkingControl\7CP7N\Project.toml`

julia> Pkg.instantiate()

(QPWalkingControl) pkg> status
Project QPWalkingControl v0.1.0
    Status `C:\Users\mark_\.julia\packages\QPWalkingControl\7CP7N\Project.toml`
  [436ed74b] AtlasRobot v0.4.0
  [ad839575] Blink v0.10.1
  [150eb455] CoordinateTransformations v0.5.0
  [4d00f742] GeometryTypes v0.7.5
  [c601a237] Interact v0.10.2
  [d3863d7c] InteractBase v0.10.2
  [9c8b4983] LightXML v0.8.0
  [b8f27783] MathOptInterface v0.8.4
  [931e9471] MechanismGeometries v0.3.0
  [283c5d60] MeshCat v0.7.0
  [6ad125db] MeshCatMechanisms v0.4.0
  [51fcb6bd] NamedColors v0.2.0
  [ab2f91bb] OSQP v0.5.2
  [510215fc] Observables v0.2.3
  [41fac6d6] Parametron v0.9.1
  [145d500b] PlanarConvexHulls v0.3.0
  [d1d18edd] QPControl v0.1.0 #master (https://github.com/tkoolen/QPControl.jl.git)
  [366cf18f] RigidBodyDynamics v2.1.0
  [e61f16d8] RigidBodySim v1.3.0
  [6038ab10] Rotations v0.11.1
  [90137ffa] StaticArrays v0.11.0
  [6f584044] StaticUnivariatePolynomials v0.6.0
  [0f1e0344] WebIO v0.7.0
  [37e2e46d] LinearAlgebra
  [8dfed614] Test
```

I then executed all of the commands from the notebook up to the setanimation! command without any errors or warnings. But, on the setanimation! line I get the same error.

```julia
julia> setanimation!(vis, Animation(mvis, sol))
ERROR: MethodError: Cannot `convert` an object of type MechanismVisualizer{MechanismState{Float64,Float64,Float64,TypeSortedCollections.TypeSortedCollection{Tuple{Array{Joint{Float64,QuaternionFloating{Float64}},1},Array{Joint{Float64,Revolute{Float64}},1}},2}},Visualizer} to an object of type Dict{MeshCat.SceneTrees.Path,MeshCat.AnimationClip}

The issue would appear to be that the Manifest.toml file specifies MechCatMechanisms 0.4.0

[[MeshCatMechanisms]]
deps = ["ColorTypes", "CoordinateTransformations", "GeometryTypes", "InteractBase", "Interpolations", "LoopThrottle", "MechanismGeometries", "MeshCat", "OrderedCollections", "RigidBodyDynamics"]
git-tree-sha1 = "efab8e156448f388bd5ff02c60308179dd8b753c"
uuid = "6ad125db-dd91-5488-b820-c1df6aab299d"
version = "0.4.0"

 which doesn't include the Animation constructor added via your pull request:

function MeshCat.Animation(mvis::MechanismVisualizer,
        times::AbstractVector{<:Real},
        configurations::AbstractVector{<:AbstractVector{<:Real}};
        fps::Integer=30)

julia> methods(Animation,)
# 5 methods for generic function "(::Type)":
[1] Animation() in MeshCat at C:\Users\mark_\.julia\dev\MeshCat\src\animations.jl:54
[2] Animation(clips::Dict{MeshCat.SceneTrees.Path,MeshCat.AnimationClip}, default_framerate::Int64) in MeshCat at C:\Users\mark_\.julia\dev\MeshCat\src\animations.jl:50
[3] Animation(fps::Int64) in MeshCat at C:\Users\mark_\.julia\dev\MeshCat\src\animations.jl:54
[4] Animation(mvis::MechanismVisualizer, times::AbstractArray{#s24,1} where #s24<:Real, configurations::AbstractArray{#s15,1} where #s15<:(AbstractArray{#s14,1} where #s14<:Real); fps) in MeshCatMechanisms at C:\Users\mark_\.julia\dev\MeshCatMechanisms\src\animate.jl:37
[5] Animation(clips, default_framerate) in MeshCat at C:\Users\mark_\.julia\dev\MeshCat\src\animations.jl:50
```

Next I tried using pkg> develop to clone the latest MeshCatMechanisms with your animation-ctor branch:

```bash
C:\Users\mark_\.julia\dev\MeshCatMechanisms>git branch
  master
* tk/animation-ctor
```

after pkg> build, I tried all of the steps again, but still hit the same error. I'm don't have enough experience with Julia even see how the changes in this pull request would allow for the Animation(mvis, sol) constructor to work. I only see that you added the new constructor with 3 arguments required (shown above).

I truly appreciate your help on this, and obviously your thesis comes first. Thanks again for all of your efforts.

**Comment by tkoolen** [comment]
OK, this should be fixed by #5. Sorry about that.

In the future, could you use [code fences](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code) when posting code; otherwise you'll inadvertently use all kinds of strange Markdown formatting and ping random users when posting macros. I.e., 

````
```julia
code here
```
````

I edited your previous comments to do this.

**Comment by Mark-Bowers** [comment]
Works perfectly. Thank you!
